File: Executor/IndexSearchExecutor.cpp
#include "IndexSearchExecutor.h"

#include "IndexContext.h"

#include <algorithm>
#include <cassert>
#include <cmath>

namespace {

QueryCompileModeParameters g_ScoringParameters = kWeakAndBigramParameters;

float DocDataPrior(const DocDataEntry& entry)
{
    const float docLength = static_cast<float>(std::max<uint32_t>(1, entry.DDE_DocLength));
    static constexpr float TargetLogLength = 6.0f;
    static constexpr float Width = 4.0f;
    static constexpr float Weight = 0.15f;
    const float distance = std::abs(std::log2(docLength) - TargetLogLength);
    const float lengthQuality = std::max(0.0f, 1.0f - distance / Width);
    return Weight * lengthQuality
        + 0.10f * entry.DDE_QualityScore
        + 0.05f * entry.DDE_AuthorityScore
        - 0.10f * entry.DDE_SpamScore;
}

float DocDataScore(const DocDataEntry& entry)
{
    // Fitted formula from the 2026-06-30 full-feature GPU sweep:
    // score = 1.8*weak + 0.1*dump_bigram + 1.0*static + 0.0*prior
    //       + 1.0*quality + 0.5*authority - 2.0*spam + 0.0*both + 0.0*bigram_only.
    // Scheme A folds stream-side terms into leaf span weights:
    // unigram span weight = 1.8; bigram span weight = 0.1 * dump_bigram_span_weight(2.0) = 0.2.
    return g_ScoringParameters.QMP_StaticWeight * entry.DDE_StaticRank
        + g_ScoringParameters.QMP_PriorWeight * DocDataPrior(entry)
        + g_ScoringParameters.QMP_QualityWeight * entry.DDE_QualityScore
        + g_ScoringParameters.QMP_AuthorityWeight * entry.DDE_AuthorityScore
        - g_ScoringParameters.QMP_SpamPenalty * entry.DDE_SpamScore;
}

float VectorScoreFeature(const DocDataEntry& entry, const std::vector<float>* query)
{
    if (!query || query->size() != DOC_VECTOR_DIM || entry.DDE_VectorFlags == 0 || entry.DDE_VectorDim != DOC_VECTOR_DIM)
        return 0.0f;

    float dot = 0.0f;
    float nq = 0.0f;
    float nd = 0.0f;
    for (size_t i = 0; i < DOC_VECTOR_DIM; ++i) {
        const float q = (*query)[i];
        const float d = static_cast<float>(entry.DDE_VectorData[i]) / 128.0f;
        dot += q * d;
        nq += q * q;
        nd += d * d;
    }
    if (nq <= 0.0f || nd <= 0.0f)
        return 0.0f;

    const float cosine = dot / (std::sqrt(nq) * std::sqrt(nd));
    return g_ScoringParameters.QMP_CosineWeight * cosine;
}

}

IndexSearchExecutor::IndexSearchExecutor(const IndexContext* context)
    : m_Context(context)
{}

void IndexSearchExecutor::SetScoringParameters(const QueryCompileModeParameters& parameters)
{
    g_ScoringParameters = parameters;
}

std::vector<SearchResult> IndexSearchExecutor::Execute(std::shared_ptr<IndexReader> reader,
                                                       int topK,
                                                       const std::vector<float>* vectorQuery)
{
    if (!reader || reader->IsEnd())
        return {};

    std::vector<SearchResult> results;
    const bool boundedTopK = topK > 0;
    const size_t heapLimit = boundedTopK ? static_cast<size_t>(topK) : 0;
    if (boundedTopK)
        results.reserve(heapLimit);

    auto worseScoreFirst = [](const SearchResult& a, const SearchResult& b) {
        return a.score > b.score;
    };

    float score = 0.0;

    while (!reader->IsEnd()) {
        uint64_t docId     = reader->GetDocumentID();

        const DocDataEntry* entry = m_Context->GetDocDataEntry(docId);

        score = reader->GetScore(entry) + DocDataScore(*entry) + VectorScoreFeature(*entry, vectorQuery);

        SearchResult result{MakeReaderDocumentID(docId, reader->GetSourceMask()), score, ""};
        
        if (!boundedTopK) {
            results.push_back(std::move(result));
        } else if (results.size() < heapLimit) {
            results.push_back(std::move(result));
            std::push_heap(results.begin(), results.end(), worseScoreFirst);
        } else if (score > results.front().score) {
            std::pop_heap(results.begin(), results.end(), worseScoreFirst);
            results.back() = std::move(result);
            std::push_heap(results.begin(), results.end(), worseScoreFirst);
        }
        reader->GoNext();
    }

    SortAndTruncate(results, topK);
    return results;
}

std::vector<SearchResult> IndexSearchExecutor::ExecuteBounded(
        std::shared_ptr<IndexReader> reader,
        int topK,
        uint64_t maxVisitedDocs,
        const std::vector<float>* vectorQuery)
{
    if (!reader || reader->IsEnd() || maxVisitedDocs == 0)
        return {};

    std::vector<SearchResult> results;
    const bool boundedTopK = topK > 0;
    const size_t heapLimit = boundedTopK ? static_cast<size_t>(topK) : 0;
    if (boundedTopK)
        results.reserve(heapLimit);

    auto worseScoreFirst = [](const SearchResult& a, const SearchResult& b) {
        return a.score > b.score;
    };

    uint64_t visited = 0;
    while (!reader->IsEnd() && visited < maxVisitedDocs) {
        uint64_t docId     = reader->GetDocumentID();
        assert(m_Context);
        const DocDataEntry* entry = m_Context->GetDocDataEntry(docId);
        assert(entry);
        float    score     = reader->GetScore(entry) + DocDataScore(*entry) + VectorScoreFeature(*entry, vectorQuery);

        SearchResult result{MakeReaderDocumentID(docId, reader->GetSourceMask()), score, ""};
        if (!boundedTopK) {
            results.push_back(std::move(result));
        } else if (results.size() < heapLimit) {
            results.push_back(std::move(result));
            std::push_heap(results.begin(), results.end(), worseScoreFirst);
        } else if (score > results.front().score) {
            std::pop_heap(results.begin(), results.end(), worseScoreFirst);
            results.back() = std::move(result);
            std::push_heap(results.begin(), results.end(), worseScoreFirst);
        }
        ++visited;
        reader->GoNext();
    }

    SortAndTruncate(results, topK);
    return results;
}

std::vector<SearchResult> IndexSearchExecutor::Execute(IndexReader* reader,
                                                       int topK,
                                                       const std::vector<float>* vectorQuery)
{
    return Execute(std::shared_ptr<IndexReader>(reader, [](IndexReader*){}), topK, vectorQuery);
}

void IndexSearchExecutor::SortAndTruncate(std::vector<SearchResult>& results, int topK)
{
    std::sort(results.begin(), results.end(),
        [](const SearchResult& a, const SearchResult& b){
            return a.score > b.score;
        });

    if (topK > 0 && (int)results.size() > topK)
        results.resize(static_cast<size_t>(topK));
}

File: RustBlade/src/Executor/IndexSearchExecutor.rs
//! Direct translation of the C++ search executor; symbol names stay aligned for debugging.
#![allow(non_snake_case, non_upper_case_globals)]

use std::cmp::{Ordering, Reverse};
use std::collections::BinaryHeap;
use std::sync::{OnceLock, RwLock};

use crate::block_table::{DocDataDecodeScore, DocDataEntry, DOC_VECTOR_DIM};
use crate::eval_expression::{kWeakAndBigramParameters, QueryCompileModeParameters};
use crate::index_reader::{IndexReader, MakeReaderDocumentID};
use crate::search_result::SearchResult;

pub trait SearchExecutionContext {
    fn GetDocDataEntry(&self, docId: u64) -> Option<&DocDataEntry>;
}

#[derive(Clone, Copy, PartialEq)]
struct ScoreKey(f32);
impl Eq for ScoreKey {}
impl PartialOrd for ScoreKey {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for ScoreKey {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0.partial_cmp(&other.0).unwrap_or(Ordering::Equal)
    }
}

struct HeapEntry {
    score: ScoreKey,
    result: SearchResult,
}
impl PartialEq for HeapEntry {
    fn eq(&self, other: &Self) -> bool {
        self.score == other.score
    }
}
impl Eq for HeapEntry {}
impl PartialOrd for HeapEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
impl Ord for HeapEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        self.score.cmp(&other.score)
    }
}

pub struct IndexSearchExecutor<'a> {
    m_Context: &'a dyn SearchExecutionContext,
}

#[allow(non_snake_case)]
impl<'a> IndexSearchExecutor<'a> {
    pub fn new(context: &'a dyn SearchExecutionContext) -> Self {
        Self { m_Context: context }
    }

    pub fn SetScoringParameters(parameters: QueryCompileModeParameters) {
        *ScoringParameters().write().unwrap() = parameters;
    }

    pub fn Execute(&self, reader: &mut dyn IndexReader, topK: i32) -> Vec<SearchResult> {
        self.ExecuteWithVector(reader, topK, None)
    }

    pub fn ExecuteWithVector(
        &self,
        reader: &mut dyn IndexReader,
        topK: i32,
        vectorQuery: Option<&[f32]>,
    ) -> Vec<SearchResult> {
        if reader.IsEnd() {
            return Vec::new();
        }

        let limit = Self::TopKLimit(topK);
        let parameters = *ScoringParameters().read().unwrap();
        let mut results = Vec::new();
        let mut heap: BinaryHeap<Reverse<HeapEntry>> = BinaryHeap::new();
        while !reader.IsEnd() {
            let docId = reader.GetDocumentID();
            let entry = self
                .m_Context
                .GetDocDataEntry(docId)
                .expect("IndexReader returned a document without DocDataEntry");
            let score = reader.GetScore(entry)
                + DocDataScore(entry, &parameters)
                + VectorScoreFeature(entry, vectorQuery, &parameters);
            let result = SearchResult {
                doc_id: MakeReaderDocumentID(docId, reader.GetSourceMask()),
                score,
                snippet: String::new(),
            };

            if limit.is_none() {
                results.push(result);
            } else if heap.len() < limit.unwrap() {
                heap.push(Reverse(HeapEntry {
                    score: ScoreKey(score),
                    result,
                }));
            } else if score > heap.peek().unwrap().0.result.score {
                heap.pop();
                heap.push(Reverse(HeapEntry {
                    score: ScoreKey(score),
                    result,
                }));
            }

            reader.GoNext();
        }

        if limit.is_some() {
            results.extend(heap.into_iter().map(|Reverse(entry)| entry.result));
        }
        Self::SortAndTruncate(&mut results, limit);
        results
    }

    pub fn ExecuteBounded(
        &self,
        reader: &mut dyn IndexReader,
        topK: i32,
        maxVisitedDocs: u64,
        vectorQuery: Option<&[f32]>,
    ) -> Vec<SearchResult> {
        if reader.IsEnd() || maxVisitedDocs == 0 {
            return Vec::new();
        }

        let limit = Self::TopKLimit(topK);
        let parameters = *ScoringParameters().read().unwrap();
        let mut results = Vec::new();
        let mut heap: BinaryHeap<Reverse<HeapEntry>> = BinaryHeap::new();
        let mut visited = 0u64;
        while !reader.IsEnd() && visited < maxVisitedDocs {
            let docId = reader.GetDocumentID();
            let entry = self
                .m_Context
                .GetDocDataEntry(docId)
                .expect("IndexReader returned a document without DocDataEntry");
            let score = reader.GetScore(entry)
                + DocDataScore(entry, &parameters)
                + VectorScoreFeature(entry, vectorQuery, &parameters);
            let result = SearchResult {
                doc_id: MakeReaderDocumentID(docId, reader.GetSourceMask()),
                score,
                snippet: String::new(),
            };

            if limit.is_none() {
                results.push(result);
            } else if heap.len() < limit.unwrap() {
                heap.push(Reverse(HeapEntry {
                    score: ScoreKey(score),
                    result,
                }));
            } else if score > heap.peek().unwrap().0.result.score {
                heap.pop();
                heap.push(Reverse(HeapEntry {
                    score: ScoreKey(score),
                    result,
                }));
            }

            visited += 1;
            reader.GoNext();
        }

        if limit.is_some() {
            results.extend(heap.into_iter().map(|Reverse(entry)| entry.result));
        }
        Self::SortAndTruncate(&mut results, limit);
        results
    }

    fn TopKLimit(topK: i32) -> Option<usize> {
        usize::try_from(topK).ok().filter(|limit| *limit > 0)
    }

    fn SortAndTruncate(results: &mut Vec<SearchResult>, limit: Option<usize>) {
        results.sort_by(|a, b| {
            if a.score > b.score {
                Ordering::Less
            } else if b.score > a.score {
                Ordering::Greater
            } else {
                Ordering::Equal
            }
        });
        if let Some(limit) = limit {
            if results.len() > limit {
                results.truncate(limit);
            }
        }
    }
}

#[allow(non_snake_case)]
fn ScoringParameters() -> &'static RwLock<QueryCompileModeParameters> {
    static PARAMETERS: OnceLock<RwLock<QueryCompileModeParameters>> = OnceLock::new();
    PARAMETERS.get_or_init(|| RwLock::new(kWeakAndBigramParameters))
}

#[allow(non_snake_case)]
fn DocDataPrior(entry: &DocDataEntry) -> f32 {
    let docLength = entry.DDE_BodyLength.max(1) as f32;
    let lengthQuality = (1.0 - (docLength.log2() - 6.0).abs() / 4.0).max(0.0);
    0.15 * lengthQuality
        + 0.10 * DocDataDecodeScore(entry.DDE_QualityScore)
        + 0.05 * DocDataDecodeScore(entry.DDE_AuthorityScore)
        - 0.10 * DocDataDecodeScore(entry.DDE_SpamScore)
}

#[allow(non_snake_case)]
fn DocDataScore(entry: &DocDataEntry, parameters: &QueryCompileModeParameters) -> f32 {
    parameters.QMP_StaticWeight * DocDataDecodeScore(entry.DDE_StaticRank)
        + parameters.QMP_PriorWeight * DocDataPrior(entry)
        + parameters.QMP_QualityWeight * DocDataDecodeScore(entry.DDE_QualityScore)
        + parameters.QMP_AuthorityWeight * DocDataDecodeScore(entry.DDE_AuthorityScore)
        - parameters.QMP_SpamPenalty * DocDataDecodeScore(entry.DDE_SpamScore)
}

#[allow(non_snake_case)]
fn VectorScoreFeature(
    entry: &DocDataEntry,
    query: Option<&[f32]>,
    parameters: &QueryCompileModeParameters,
) -> f32 {
    let Some(query) = query else {
        return 0.0;
    };
    if query.len() != DOC_VECTOR_DIM
        || entry.DDE_VectorDim as usize != DOC_VECTOR_DIM
        || entry.DDE_VectorFormat == 0
    {
        return 0.0;
    }

    let mut dot = 0.0f32;
    let mut nq = 0.0f32;
    let mut nd = 0.0f32;
    for i in 0..DOC_VECTOR_DIM {
        let q = query[i];
        let d = entry.DDE_VectorData[i] as f32 / 128.0;
        dot += q * d;
        nq += q * q;
        nd += d * d;
    }
    if nq <= 0.0 || nd <= 0.0 {
        return 0.0;
    }
    parameters.QMP_CosineWeight * dot / (nq.sqrt() * nd.sqrt())
}
